### PR TITLE
JDK-8326496: [test] checkHsErrFileContent support printing hserr in error case

### PR DESCRIPTION
**Pull request by MBaesken**
checkHsErrFileContent checks hs error files for existance and non-existance of some user provided patterns. However it should also have an option to print the hs_err file in case of a failing check, this would make error analysis easier.

---------
### Progress
- [x] Change must be properly reviewed (1 review required, with at least 1 [Reviewer](https://openjdk.org/bylaws#reviewer))
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue

### Issue
 * [JDK-8326496](https://bugs.openjdk.org/browse/JDK-8326496): [test] checkHsErrFileContent  support printing hserr in error case (**Enhancement** - P4)

### Reviewers
 * [Christoph Langer](https://openjdk.org/census#clanger) (@RealCLanger - **Reviewer**)
 * [Lutz Schmidt](https://openjdk.org/census#lucy) (@RealLucy - **Reviewer**)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/jdk.git pull/17978/head:pull/17978` \
`$ git checkout pull/17978`

Update a local copy of the PR: \
`$ git checkout pull/17978` \
`$ git pull https://git.openjdk.org/jdk.git pull/17978/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 17978`

View PR using the GUI difftool: \
`$ git pr show -t 17978`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/jdk/pull/17978.diff">https://git.openjdk.org/jdk/pull/17978.diff</a>

</details>

### Webrev
[Link to Webrev Comment](https://git.openjdk.org/jdk/pull/17978#issuecomment-1960989207)

### CODE REVIEW

File: test/hotspot/jtreg/runtime/ErrorHandling/HsErrFileUtils.java [hunk 1]
```diff
@@ -63,12 +63,10 @@ public static File openHsErrFileFromOutput(OutputAnalyzer output) {
      * if patterns are missing, or if the END marker is missing.
      * @param f Input file
      * @param patterns An array of patterns that need to match, in that order
-     * @param verbose If true, the content of the hs-err file is printed while matching. If false, only the matched patterns
-     *                are printed.
      * @throws RuntimeException, {@link IOException}
```
Comment (RealCLanger) [inline test/hotspot/jtreg/runtime/ErrorHandling/HsErrFileUtils.java:66]: The Javadoc of parameter "verbose" should not be removed here.

**Review by RealCLanger — APPROVED** [review]
One minor remark, looks good otherwise.

File: test/hotspot/jtreg/runtime/ErrorHandling/HsErrFileUtils.java (hunk 1 above)
Comment (MBaesken) [reply test/hotspot/jtreg/runtime/ErrorHandling/HsErrFileUtils.java:68]: okay I added it again; but it was already not documented in the same way before my change, once "the matched patterns are printed" once "only important information are printed" . Should we adjust it and to what ?

Comment (MBaesken) [reply test/hotspot/jtreg/runtime/ErrorHandling/HsErrFileUtils.java:68]: > Looking at what "verbose" does, I'd say the correct description would be "only the matched patterns are printed"

Mostly correct, but there could be also "Lazily skipping the rest of the hs-err file..."  for example; but mostly (almost all) output is about the patterns ("Matches ..." or "Forbidden pattern found  ..." ).

**Review by RealCLanger — APPROVED** [review]

**Review by dholmes-ora — COMMENTED** [review]
Sorry I'm struggling to understand under exactly what conditions the hs_err file will now be printed. Can you give some concrete examples, and why you want to print a potentially very large file as part of the tests main output (which may overflow the jtreg output buffer). Thanks

**Comment by MBaesken** [comment]
> Sorry I'm struggling to understand under exactly what conditions the hs_err file will now be printed. 

It will be printed in case the  checkHsErrFileContent  fails,  So we do not have to set 'verbose' and print the content always (because in the successful case it is not really needed) but only in failure case.

> Can you give some concrete examples, and why you want to print a potentially very large file as part of the tests main output (which may overflow the jtreg output buffer). Thanks

I want to print it to see what was wrong with the hserr file content.  Especially in the SecondaryErrorTest we had errors where we missed the hserr content to look WHAT exactly was really bad/failing.

**Comment by dholmes-ora** [comment]
Sorry but when you see an error like:
```
Forbidden pattern found at line " + lineNo + ": " + negativePattern)
```
why does that not suffice? And if it doesn't suffice why don't you just grab the hs_err file from the test artifacts and examine it directly?

**Comment by RealCLanger** [comment]
Hi David,

we sometimes see error output like this:

[2024-02-09T01:23:06.112705412Z] Gathering output for process 47059
Will crash now (TestCrashInErrorHandler=14)...
^^^ Matches Will crash now \\(TestCrashInErrorHandler=14\\)... at line 2^^^
[error occurred during error reporting (test secondary crash 1), id 0xb, SIGSEGV (0xb) at pc=0x000074d8362e1664]
^^^ Matches \\[error occurred during error reporting \\(test secondary crash 1\\).*\\] at line 4^^^
[siginfo: si_signo: 11 (SIGSEGV), si_code: 1 (SEGV_MAPERR), si_addr: 0x0000000000000400]
^^^ Matches \\[siginfo:.*\\(SIGSEGV\\).*\\] at line 5^^^
[stack: Native frames: (J=compiled Java code, j=interpreted, Vv=VM code, C=native code)
^^^ Matches \\[stack: Native frames:.* at line 6^^^
V  [libjvm.so+0x1d01664]  VMError::controlled_crash(int)+0x194
^^^ Matches .*VMError::controlled_crash.* at line 7^^^
Will crash now (TestCrashInErrorHandler=14)...
^^^ Matches Will crash now \\(TestCrashInErrorHandler=14\\)... at line 9^^^
[error occurred during error reporting (test secondary crash 2), id 0xb, SIGSEGV (0xb) at pc=0x000074d8362e1664]
^^^ Matches \\[error occurred during error reporting \\(test secondary crash 2\\).*\\] at line 11^^^
----------System.err:(12/723)*----------
java.lang.RuntimeException: hs-err file incomplete (first missing pattern: \\[siginfo:.*\\(SIGSEGV\\).*\\])
	at HsErrFileUtils.checkHsErrFileContent(HsErrFileUtils.java:135)
	at HsErrFileUtils.checkHsErrFileContent(HsErrFileUtils.java:71)
	at SecondaryErrorTest.main(SecondaryErrorTest.java:125)
	at java.base/jdk.internal.reflect.DirectMethodHandleAccessor.invoke(DirectMethodHandleAccessor.java:103)
	at java.base/java.lang.reflect.Method.invoke(Method.java:580)
	at com.sun.javatest.regtest.agent.MainActionHelper$AgentVMRunnable.run(MainActionHelper.java:333)
	at java.base/java.lang.Thread.run(Thread.java:1570)

So in that case (pattern missing) I'd really like to have the full hs_err file to see what happened.

But I agree that for things like Forbidden pattern found, as you mentioned above, one would not need to print the whole hs_err file. Maybe the change could be checked/modified to only print the hs_err content if something is missing?

**Comment by MBaesken** [comment]
Hi David,

>why does that not suffice? 
>And if it doesn't suffice why don't you just grab the hs_err file from the test artifacts and examine it directly?

For the forbidden patterns it might indeed suffice (however having more information understanding the context of why a forbidden pattern occurs is still often necessary).
On some shielded containers / Jenkins setups I usually only have the log and cannot 'collect" whatever information.  This is not the exception but a very common thing.
(In my local environment I do not see the error , so no hserr is printed)
Btw. we are totally fine with a _verbose_ parameter of the check function (prints the hserr content even in case of success) .

**Comment by dholmes-ora** [comment]
I remain concerned that printing the hs_err file within the log file may actually cause more problems than it fixes. Test environments should gather up all test artifacts for examination upon failure so the actual hs_err file should be available IMO. But if you think you need this ...

**Review by RealLucy — APPROVED** [review]
Looks good.

**Pull request closed without merging**